proposals/compile-time-dependency-resolution.md

```diff
@@ -0,0 +1,590 @@
+
+#### 1. Arguments of the caller function
+
+It looks into the caller function argument list for an evidence of the required extension. Here, `bothValid()` gets a `Validator<A>` passed in so, whenever it needs to resolve it for the inner calls to `validate()`, it'll be able to retrieve it from its own argument list.
+```kotlin
+fun <A> validate(a: A, with validator: Validator<A>): Boolean = a.isValid()
+
+fun <A> bothValid(x: A, y: A, with validator: Validator<A>): Boolean = validate(x) && validate(y)
+```
+
+#### 2. Companion object for the target type
+
+In case there's no evidence at the caller function level, we'll look into the companion of the target type. Let's say we have an extension of `Validator<User>`:
+```kotlin
+package com.domain
+
+data class User(val id: Int, val name: String) {
+  companion object {
+    extension class UserValidator(): Validator<User> {
+      override fun User.isValid(): Boolean {
+        return id > 0 && name.length > 0
+      }
+    }
+  }
+}
+```
+That'll be enough for resolving the extension.
+
+#### 3. Companion object for the constraint interface we're looking for
+
+In case neither evidence exists in the companion of the target type, we'll look in the companion of the constraint interface:
+
+```kotlin
+interface Validator<A> {
+  fun A.isValid(): Boolean
+
+  companion object {
+    extension class GroupValidator<A>(with val userValidator: Validator<User>) : Validator<Group> {
+      override fun Group.isValid(): Boolean {
+        for (x in users) {
+          if (!x.isValid()) return false
+        }
+        return true
+      }
+    }
+  }
+}
+```
+
+#### 4. Subpackages of the package where the target type is defined
+
+The next step would be to look into the subpackages of the package where the target type (`User`) is declared. It will simply look in subpackages under the current gradle module, it doesn't support cross-module definitions. These extensions **must be flagged as `internal`**.
+
+```kotlin
+package com.domain.repository
+
+import com.domain.User
+
+internal extension object UserRepository : Repository<User> {
+
+  val storedUsers: MutableMap<Int, User> = mutableMapOf() // e.g. users stored in a DB
+
+  override fun loadAll(): Map<Int, User> {
+    return storedUsers
+  }
+
+  override fun loadById(id: Int): User? {
+    return storedUsers[id]
+  }
+
+  override fun User.save() {
+    storedUsers[this.id] = this
+  }
+}
+```
+
+Here we got a `Repository<User>` defined in a subpackage of `com.domain`, where the `User` type is defined.
+
+#### 5. Subpackages of the package where the constraint interface is defined
+
+The last place to look would be the subpackages of the package where the constraint interface is defined. t will simply look in subpackages under the current gradle module, it doesn't support cross-module definitions. These extensions **must be flagged as `internal`**.
```

Comment:
missing `I`:
`defined. t will simply look in subpackages`

Reply:
```suggestion
The last place to look would be the subpackages of the package where the constraint interface is defined. It will simply look in subpackages under the current gradle module, it doesn't support cross-module definitions. These extensions **must be flagged as `internal`**.
```